json-core/src/main/java/io/micronaut/json/JsonCodec.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2017-2021 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.json;
+
+import io.micronaut.core.annotation.AnnotationMetadata;
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.convert.ConversionError;
+import io.micronaut.core.convert.exceptions.ConversionErrorException;
+import io.micronaut.core.type.Argument;
+import io.micronaut.json.tree.JsonNode;
+import org.reactivestreams.Processor;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+import java.util.function.Consumer;
+
+public interface JsonCodec {
```

Comment:
I think naming this `JsonCodec` is just going to confuse matters with our existing codec API so maybe a different name is needed here like `JsonMapper` or whatever.

Eitherway this is new public API and needs javadoc for the type and all public methods that are not marked with `@Internal`

Each method parameter should also be annotated with `@Nullable` or `@NotNull` to express nullability which helps kotlin users  as well